File: backend/app/graphql/mutations/purchase_points_product.rb
module Mutations
  class PurchasePointsProduct < Mutations::BaseMutation
    description "Purchase a points product using customer points"

    # Define the arguments (inputs) for this mutation
    argument :product_id, String, required: true, description: "ID of the points product to purchase"
    argument :customer_id, String, required: true, description: "ID of the customer making the purchase"
    argument :points_amount, Integer, required: true, description: "Number of points to spend on this purchase"

    # Define the return fields
    field :success, Boolean, null: false, description: "Whether the purchase was successful"
    field :message, String, null: true, description: "Success or error message"
    field :reward, Types::RewardType, null: true, description: "The reward that was earned"

    # Define the resolve method
    def resolve(product_id:, customer_id:, points_amount:)
      client = ::Api::Smile.new(api_key: Rails.application.credentials.smile.private_key)
      response = client.purchase_points_product(product_id, customer_id, points_amount)
      {
        success: true,
        message: "Successfully purchased points product",
        reward: response['points_purchase']['reward_fulfillment']
      }
    rescue StandardError => e
      # Handle errors from the API call
      {
        success: false,
        message: "Error: #{e.message}",
        reward: nil
      }
    end
  end
end 

File: frontend/src/client/api/smile/purchase-points-product.ts
import { Reward } from "./types"
import getConfig from "../../config"

const PURCHASE_POINTS_PRODUCT_MUTATION = `
  mutation PurchasePointsProduct($input: PurchasePointsProductInput!) {
    purchasePointsProduct(input: $input) {
      success
      message
      reward {
        id
        name
        code
        imageUrl
        actionText
        actionUrl
        usageInstructions
        termsAndConditions
        expiresAt
      }
    }
  }
`;

const purchasePointsProduct = async (productId: string, customerId: string, pointsAmount: number): Promise<Reward> => {
  const { perkUpAPI } = getConfig();
  
  try {
    const response = await fetch(`${perkUpAPI}/graphql`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: PURCHASE_POINTS_PRODUCT_MUTATION,
        variables: {
          input: {
            productId,
            customerId,
            pointsAmount
          }
        }
      })
    });

    const data = await response.json();
    
    if (data.errors) {
      console.error('Error purchasing points product:', data.errors);
      throw new Error('Error purchasing points product');
    }
  
    console.log('Purchase points product response:', data);
    return data.data.purchasePointsProduct.reward;
  } catch (error) {
    console.error('Error purchasing points product:', error);
    throw new Error('Error purchasing points product');
  }
};

export { purchasePointsProduct };